gremlin-dotnet/src/Gremlin.Net/Driver/Connection.cs

```diff
@@ -55,87 +71,86 @@ internal class Connection : IConnection
-            ResponseStatus status;
-            IAggregator aggregator = null;
-            var isAggregatingSideEffects = false;
-            var result = new List<T>();
-            do
-            {
-                var received = await _webSocketConnection.ReceiveMessageAsync().ConfigureAwait(false);
-                var receivedMsg = _messageSerializer.DeserializeMessage<ResponseMessage<JToken>>(received);
-
-                status = receivedMsg.Status;
-                status.ThrowIfStatusIndicatesError();
+            var state = Volatile.Read(ref _connectionState);
+            if (state == Closed) return;
+            ReceiveMessagesAsync().Forget();
+        }
 
-                if (status.Code == ResponseStatusCode.Authenticate)
+        private async Task ReceiveMessagesAsync()
+        {
+            while (true)
+            {
+                try
                 {
-                    await AuthenticateAsync().ConfigureAwait(false);
+                    var received = await _webSocketConnection.ReceiveMessageAsync().ConfigureAwait(false);
+                    Parse(received);
                 }
-                else if (status.Code != ResponseStatusCode.NoContent)
+                catch (Exception e)
                 {
-                    var receivedData = typeof(T) == typeof(JToken)
-                        ? new[] { receivedMsg.Result.Data }
-                        : _graphSONReader.ToObject(receivedMsg.Result.Data);
-
-                    foreach (var d in receivedData)
-                        if (receivedMsg.Result.Meta.ContainsKey(Tokens.ArgsSideEffectKey))
-                        {
-                            if (aggregator == null)
-                                aggregator =
-                                    new AggregatorFactory().GetAggregatorFor(
-                                        (string) receivedMsg.Result.Meta[Tokens.ArgsAggregateTo]);
-                            aggregator.Add(d);
-                            isAggregatingSideEffects = true;
-                        }
-                        else
-                        {
-                            result.Add(d);
-                        }
+                    await CloseConnectionBecauseOfFailureAsync(e).ConfigureAwait(false);
+                    break;
                 }
+            }
+        }
 
-                if (status.Code == ResponseStatusCode.Success || status.Code == ResponseStatusCode.NoContent)
-                {
-                    statusAttributes = receivedMsg.Status.Attributes;
-                }
+        private void Parse(byte[] received)
+        {
+            var receivedMsg = _messageSerializer.DeserializeMessage<ResponseMessage<JToken>>(received);
+            
+            try
+            {
+                TryParseResponseMessage(receivedMsg);
+            }
+            catch (Exception e)
+            {
+                _callbackByRequestId[receivedMsg.RequestId].HandleFailure(e);
```

Comment:
NIT, we can avoid the second read access to the dictionary by using:

```C#
if (_callbackByRequestId.TryRemove(receivedMsg.RequestId, out var responseHandler))
{
    responseHandler.HandleFailure(e);
}
```

Reply:
Oh yes, kind of obvious. I just pushed a commit with this change.